### DISCUSSION TITLE
EMCEE walkers not exploring parameter space

### DISCUSSION THREAD

**Discussion by rfreire-400** [phoebe-project/phoebe2#767] [Questions]
I'm trying to use the EMCEE sampler but I'm getting some weird results. The lnprobability plot isn't exploring the parameter space and remains constant at a very low probability.
![image](https://github.com/phoebe-project/phoebe2/assets/113457579/e60a0d4c-016d-4962-9bb8-163135eb10e2)
I checked the trace plots and corner plot and they appear to be working as expected:
![image](https://github.com/phoebe-project/phoebe2/assets/113457579/52ffdd83-baa3-43f8-a16c-88f55fdbc676)
![image](https://github.com/phoebe-project/phoebe2/assets/113457579/836082d2-cd5a-4cce-b169-6e2503dd3cdb)

The results from the corner plot don't match existing literature of the system, but I'm assuming that it's normal since I didn't run the sampler for very long. I tried running it for 5000 iterations instead. While the corner and trace plots would behave like the ones above, the lnprobability plot changed a bit.

![image](https://github.com/phoebe-project/phoebe2/assets/113457579/0ef59dc3-fb2a-4da6-9ecb-039a175bf65b)

I tried troubleshooting this by running it with less parameters, changing the priors, increasing the range of the priors, running the sampler with/without flipped constraints... but the lnprobability plots were always constant at a negative probability. I would usually run the code using ellc to make it run faster. I tried running it without ellc but it's much slower and I don't have any results for that yet. Here's how I would initialize my distributions:

```
b.add_distribution({'sma@binary': phoebe.gaussian_around(0.1),
                    'incl@binary': phoebe.gaussian_around(2.0),
                    'q@binary': phoebe.gaussian_around(0.1),
                    'distance@system': phoebe.gaussian_around(160),
                    'teffratio@binary': phoebe.gaussian_around(0.05),
                    'requivratio@binary': phoebe.gaussian_around(0.05),
                    'requivsumfrac@binary': phoebe.gaussian_around(0.025),
                    'teff@primary': phoebe.gaussian_around(100),
                    'gravb_bol@primary': phoebe.uniform(low=0.9, high=1.0),
                    'gravb_bol@secondary': phoebe.uniform(low=0.32, high=1.0),
                    'irrad_frac_refl_bol@primary': phoebe.uniform(low=0.9, high=1.0),
                    'irrad_frac_refl_bol@secondary': phoebe.uniform(low=0.6, high=1.0),
                    'esinw@binary': phoebe.gaussian_around(0.01),
                    'ecosw@binary': phoebe.gaussian_around(0.01),
                    'sigmas_lnf@lc01': phoebe.gaussian_around(1)
                    }, distribution='ball_around_guess', overwrite_all=True)

b.set_value('init_from', 'ball_around_guess')

b.add_distribution({'sma@binary': phoebe.uniform(low=1.0, high=60),
                    'incl@binary': phoebe.uniform(low=0.0, high=90),
                    'teff@primary': phoebe.uniform(low=3000, high=25000),
                    'teff@secondary': phoebe.uniform(low=3000, high=25000),
                    'requiv@primary': phoebe.uniform(low=0.1, high=15),
                    'requiv@secondary': phoebe.uniform(low=0.1, high=15),
                    'q@binary': phoebe.uniform(low=0.0, high=1.0)
                    }, distribution='priors_from_external_source', overwrite_all=True )
b.set_value('priors', 'priors_from_external_source')

b.set_value('nwalkers', solver='emcee_solver', value=30)
b.set_value('niters', solver='emcee_solver', value=2500)

b.run_solver('emcee_solver', solution='emcee_sol')
```
Does anyone know why it's doing this? Is it possible that it's stuck on some local minimum?

**Comment by kecnry** [comment]
From the trace plots, it doesn't seem that the walkers themselves are stuck, but I am confused why it appears that lnprobabilities aren't affected by the changes in parameter values.  Are the lnprobabilities truly identical or just very similar (by inspecting the array in `b.get_value('lnprobabilities', solution='emcee_sol')`?

Can you also try sampling the forward model over the initial distribution and ensure that the changes are resulting in differences to the synthetic model itself (maybe ellc is giving the same light curve back despite the change in parameter value which would explain lnprobability appearing to be stuck)?  You can do that with something like: 

```
b.run_compute(sample_from='ball_around_guess', sample_num=20, sample_mode='all', model='sampled_over_dist')
b.plot(model='sampled_over_dist', show=True)
```

**Comment by rfreire-400** [comment]
Sorry for taking so long, I'm running this on a remote cluster and sending the solutions back to my local machine which took me a while to get everything settled.

I checked if the values are truly constant and here's what I got:
<img width="641" alt="Screenshot 2023-07-13 at 9 43 45 AM" src="https://github.com/phoebe-project/phoebe2/assets/113457579/a5c3fc26-501c-4573-977a-7e7f830bbe3f">

I'm not sure if I ran the forward model properly because here's the model that I got:
![forward_model_mcmc](https://github.com/phoebe-project/phoebe2/assets/113457579/111e19c9-47e6-4469-a922-8709d0711be2)

I ran it a couple of times changing different things and got this every time. 
Was I supposed to replace 'ball_around_guess' with 'emcee_sol'?

**Reply by kecnry** [reply]
Ok, your model is nowhere near your actual data, so you'll need to account for [passband luminosity scaling](http://phoebe-project.org/docs/latest/tutorials/pblum.ipynb), [distance](http://phoebe-project.org/docs/latest/tutorials/distance.ipynb), and/or [third light](http://phoebe-project.org/docs/latest/tutorials/l3.ipynb).

To make sure the forward model is still changing when the parameters change, can you plot without the data?

```
b.plot(model='sampled_over_dist', context='model', show=True)
```

**Reply by rfreire-400** [reply]
Here's the plot without the data.
![forward_model2_mcmc](https://github.com/phoebe-project/phoebe2/assets/113457579/d227b85e-5a38-40b5-8066-7eb3e281807c)

I'm getting my data in electrons / second and then normalizing it before sending it to Phoebe. I usually run  plum_mode with dataset-scaled so I don't have to do the conversion to watts / m^2 for now

**Reply by kecnry** [reply]
Ok, it's good to see the model is varying, so at least the problem isn't in the ellc wrapper itself.  I think the huge discrepancy in scales (75 orders of magnitude) between the synthetic model and observations is what is driving the lnprobabilities to be so extreme (they may not in fact be identical after 9 digits, but those first 9 digits might be dominated by this offset of 75 orders).

Is it pblum_mode to dataset-scaled now?  If so, maybe there is a bug in the dataset scaling as it doesn't seem to be working when calling run_compute.  And if not, then you should either set it or adjust the other scaling options and make sure the synthetic model drawn from your distributions and roughly agree with your observations.

**Reply by rfreire-400** [reply]
Pblum_mode wasn't set to dataset-scaled as I was getting some warnings suggesting I change it. I reran it with dataset-scaled now and there is some improvement although it's not perfect.
![requivfrac+ratio_trace](https://github.com/phoebe-project/phoebe2/assets/113457579/29d8296b-9666-4942-bbc8-07a020d65274)
![requivfrac+ratio_lnprob](https://github.com/phoebe-project/phoebe2/assets/113457579/86a076a8-5e57-4b6d-a9ce-463dcd3c8250)
![requivfrac+ratio_corner](https://github.com/phoebe-project/phoebe2/assets/113457579/08663083-0392-41cc-a3ca-0ca4c6eb67b6)

The lnprobability plot is no longer constant but it's seems that all of the walkers are stuck around the same solution every time

**Reply by kecnry** [reply]
Great - it looks to be working as expected!  The lnprobability is still climbing significantly, even after 1000 iterations - so my suggestion would be to use optimizers first and really spend time constructing the initializing distributions so that you do not waste a lot of computation time during the burnin phase.

**Reply by rfreire-400** [reply]
Will do, thanks for your help!
